### ISSUE TITLE
Typescript source/definitions issues

### ISSUE THREAD

**Issue by nmocruz** [Profiscience/knockout-contrib#104]
Hi, 

I was trying to include this library in a typescript project, but maybe I'm missing something in my tsconfig.json because it starts to check source code in this module inside node_modules even excluding "node_modules". maybe is a noobie question about typescript, or it can be done something to improve here, because it tries to check ts files under node_modules/@profiscience/knockout-contrib-* using my tsconfig.json.

I'm using webpack and awesome-typescript-loader, maybe is required to do something here?

**Comment by caseyWebb** [comment]
tl;dr: Add `esnext.asynciterable` to the `lib` in your tsconfig, and enable `allowSyntheticDefaultImports`.

Please don't close this issue. I'll do it when this is added to the documentation.

---

__Explanation:__

So, in an ideal world, when publishing a library, you build definition files (.d.ts), and then when consumers of that library use it with tsc, it will use those to type-check their code.

I've tried to get these working in the past, but faced issues.

The first is that in order to use definitions, the `types` property in the package.json files for individual packages must be pointed to the built definitions. Because this is a monorepo, it leads to the problem where you have to build packages in such an order that packages that are depended on by other packages are built prior to the packages that depend on them, so that those definitions are built.

The second, even more glaring issue, is that tsc blows up when you try to build definitions for mixins, for whatever reason. 

Because of this, the `types` property is pointed to the source code. This means, that in order for it to type check your code that is consuming this library, it must be able to parse this library. Because of certain features used (async generators, e.g. `async function* myFunc() {}`), you've got to tell tsc that the necessary runtime features exist.

The first issue looks like it's been resolved with TS 3.0.0, although I'm not sure how the build performance will compare; if the second issue isn't resolved though it won't really matter anyway.

**Comment by nmocruz** [comment]
Ok, make sense. I think that I tried but there's more conflicts, one is
related with Jsx other is a missmatch in a type due an init function.
I think that maybe we need a new approach to publish only type definitions.
Like declare somewhere this list of packages and their dependencies and
peek free or resolved ones first.

**Comment by nmocruz** [comment]
I forked and tried to use the typescript compile API to solve that, it still needs some work.
I'm able to generate d.ts files but I need to run ts-node excluding this d.ts  as input files so they can be overwritten. 
Then is needed to update packages.jon to point to index.d.ts not to index.ts

**Comment by caseyWebb** [comment]
@nmocruz I'd definitely prefer to only publish type definitions, I just haven't been able to do it successfully without making the build take a loooooong time.

**Comment by nmocruz** [comment]
Process all that packages will take some time, looking the Compiler API
there's something related with incremental build, not sure if will speed up
or not individual packages significantly.

**Comment by caseyWebb** [comment]
@nmocruz I'm under the impression this is all working? Can you review [this documentation](https://github.com/Profiscience/knockout-contrib/blob/9292304ea6ddd66ff241e9ef9476389ed6cca986/packages/_/README.md#using-the-esnext-build) to confirm that it covers everything?
